File: Python/DFS/39. Combination Sum.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        temp = []
        def dfs(cursum, path):
            if cursum > target:
                return
            elif cursum == target:
                temp.append(path)
            for i in candidates:
                dfs(cursum + i, path+[i])
        
        dfs(0, [])
        res = []
        for l in temp:
            if sorted(l) not in res:
                res.append(sorted(l))

        return res

class Solution1:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        res = []
        def dfs(cursum, index, path):
            if cursum > target:
                return
            elif cursum == target:
                res.append(path)
            for i in range(index, len(candidates)):
                dfs(cursum + candidates[i], i, path+[candidates[i]])
        
        dfs(0, 0, [])

        return res

if __name__ == '__main__':
    candidates = [2,3,6,7]
    target = 7
    sol = Solution()
    print(sol.combinationSum(candidates, target))

File: CPP/DFS/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        ans.clear();

        vector<int> temp;

        solve(0, candidates, temp, target);

        return ans;
    }
    void solve(int start, vector<int>& candidates, vector<int>& temp, int target){
        if(target == 0){
            ans.push_back(temp);
            return;
        }

        if(target < 0)
            return;
        
        for(int i=start; i<candidates.size(); i++){
            temp.push_back(candidates[i]);
            solve(i, candidates, temp, target - candidates[i]);
            temp.pop_back();
        }

    }
};